File: python_version/Zwierze.py
from Organizm import Organizm
import random

class Zwierze(Organizm):
    def akcja(self):
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        kierunek = random.randint(0, 3)

        nowy_x = self.x + dx[kierunek]
        nowy_y = self.y + dy[kierunek]

        if 0 <= nowy_x < self.swiat.get_szerokosc() and 0 <= nowy_y < self.swiat.get_wysokosc():
            self.wykonaj_ruch(nowy_x, nowy_y)

    def kolizja(self, inny):
        if self.czy_ten_sam_gatunek(inny):
            for dx, dy in [(-1,0), (1,0), (0,-1), (0,1)]:
                nowy_x = self.x + dx
                nowy_y = self.y + dy
                if self.swiat.czy_pole_puste(nowy_x, nowy_y):
                    potomek = self.stworz_potomka(nowy_x, nowy_y)
                    self.swiat.dodaj_organizm(potomek)
                    self.swiat.dodaj_log(f"{self.get_nazwa()} rozmnaża się na polu ({nowy_x}, {nowy_y})")
                    return
        else:
            if inny.czy_odbil_atak(self):
                self.swiat.dodaj_log(f"{self.get_nazwa()} odbił się od {inny.get_nazwa()}")
                return
            elif self.sila >= inny.get_sila():
                if inny.get_nazwa() == "Guarana":
                    self.zwieksz_sile(3)
                    self.swiat.dodaj_log(f"{self.get_nazwa()} zjadł Guaranę i zyskał 3 siły.")
                else:
                    self.swiat.dodaj_log(f"{self.get_nazwa()} zabił {inny.get_nazwa()}")
                inny.zabij()
            else:
                self.swiat.dodaj_log(f"{self.get_nazwa()} został zabity przez {inny.get_nazwa()}")
                self.zabij()

    def wykonaj_ruch(self, nowy_x, nowy_y):
        cel = self.swiat.get_organizm(nowy_x, nowy_y)
        if cel == self:
            self.swiat.dodaj_log("Błąd: kolizja samego ze sobą!")
            return

        if cel is None or not cel.czy_zyje():
            self.swiat.aktualizuj_pole(self.x, self.y, '.')
            self.x = nowy_x
            self.y = nowy_y
            self.swiat.aktualizuj_pole(self.x, self.y, self.rysowanie())
            self.swiat.dodaj_log(f"{self.get_nazwa()} przesunął się na ({self.x}, {self.y})")
        else:
            self.kolizja(cel)
            if self.czy_zyje():
                self.swiat.aktualizuj_pole(self.x, self.y, '.')
                self.x = nowy_x
                self.y = nowy_y
                self.swiat.aktualizuj_pole(self.x, self.y, self.rysowanie())

    def czy_odbil_atak(self, inny):
        return False

    def czy_ten_sam_gatunek(self, inny):
        raise NotImplementedError()

    def stworz_potomka(self, nowy_x, nowy_y):
        raise NotImplementedError()

File: java_version/Zwierze.java
import java.util.Random;

public abstract class Zwierze extends Organizm {
    public Zwierze(int sila, int inicjatywa, int x, int y, Swiat swiat) {
        super(sila, inicjatywa, x, y, swiat);
    }

    @Override
    public void akcja() {
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        int kierunek = new Random().nextInt(4);
        int nowyX = x + dx[kierunek];
        int nowyY = y + dy[kierunek];

        if (swiat.czyWPlanszy(nowyX, nowyY)) {
            wykonajRuch(nowyX, nowyY);
        }
    }

    @Override
    public void kolizja(Organizm inny) {
        if (czyTenSamGatunek(inny)) {
            int[] dx = {-1, 1, 0, 0};
            int[] dy = {0, 0, -1, 1};

            for (int i = 0; i < 4; i++) {
                int nowyX = x + dx[i];
                int nowyY = y + dy[i];
                if (swiat.czyWPlanszy(nowyX, nowyY) && swiat.czyPolePuste(nowyX, nowyY)) {
                    swiat.dodajOrganizm(stworzPotomka(nowyX, nowyY));
                    swiat.dodajLog(getNazwa() + " rozmnaża się na polu (" + x + ", " + y + ")");
                    return;
                }
            }
        } else {
            if (inny.czyOdbilAtak(this)) {
                swiat.dodajLog(getNazwa() + " na (" + x + ", " + y + ") odbił atak od " + inny.getNazwa());
                return;
            } else if (sila >= inny.getSila()) {
                if (inny.getNazwa().equals("Guarana")) {
                    swiat.dodajLog(getNazwa() + " na (" + inny.getX() + ", " + inny.getY() + ") zabił " + inny.getNazwa() + " i zyskał +3 siły.");
                    zwiekszSile(3);
                } else {
                    swiat.dodajLog(getNazwa() + " na (" + inny.getX() + ", " + inny.getY() + ") zabił " + inny.getNazwa());
                }
                inny.zabij();
            } else {
                swiat.dodajLog(getNazwa() + " na (" + inny.getX() + ", " + inny.getY() + ") został zabity przez " + inny.getNazwa());
                this.zabij();
                swiat.usunMartweOrganizmy();  // Potencjalne odświeżenie GUI
            }
        }
    }

    protected void wykonajRuch(int nowyX, int nowyY) {
        Organizm cel = swiat.getOrganizm(nowyX, nowyY);
        if (cel == this) {
            swiat.dodajLog("Błąd: próba kolizji samego ze sobą!");
            return;
        }

        if (cel == null || !cel.czyZyje()) {
            swiat.aktualizujPole(x, y, '.');
            x = nowyX;
            y = nowyY;
            swiat.aktualizujPole(x, y, rysowanie());
            swiat.dodajLog(getNazwa() + " przesunął się na pole (" + x + ", " + y + ")");
        } else {
            kolizja(cel);
            if (!czyZyje()) return;

            swiat.aktualizujPole(x, y, '.');
            x = nowyX;
            y = nowyY;
            swiat.aktualizujPole(x, y, rysowanie());
            swiat.dodajLog(getNazwa() + " przesunął się na pole (" + x + ", " + y + ")");
        }
    }

    public abstract boolean czyTenSamGatunek(Organizm inny);
}
